### ISSUE TITLE
Restructure: Packitize and simplify implementation buildpack interactions

### ISSUE THREAD

**Issue by joshzarrabi** [paketo-buildpacks/dotnet-core#322]
# Background

This language family buildpack is currently held together by a series of confusing `provides` and `requires` pairs and only works because there is only one order grouping, so all dependencies end up in every `build` and `launch` image.  This [restructure rfc](https://github.com/paketo-buildpacks/dotnet-core/blob/main/rfcs/0001-restructure.md) provides more background in the current situation and the direction the buildpack should go in.

# Proposal

In order to  make the buildpack conform to the above RFC, we need to 

* Packtize every implementation buildpack
* Rework the producer buildpacks to only `provide` thier dependency and `require` nothing.
* Rework `dotnet-core-conf` and `dotnet-core-build` to only require the dependencies they need when they need them.

Below is a proposed way to accomplish those goals while keeping the language family useable.

### Step one -- Rename and packitize `dotnet-core-conf` and `dotnet-core-build`

These buildpacks should be renamed to be more descriptive. and need to be rewritten with `packit`. For now all functionality should be left the same.

- [x] https://github.com/paketo-buildpacks/dotnet-core-conf/issues/77
- [x] https://github.com/paketo-buildpacks/dotnet-core-build/issues/84

Also, the `icu` buildpack needs to be rewritten with `packit`, everything else about it is fine though.

- [x] https://github.com/paketo-buildpacks/icu/issues/67

### Step 2 -- Rework the `requires` for `dotnet-core-publish` and `dotnet-core-execute`

Currently, the logic for getting the right versions of the right dependencies into the correct image for these buildpacks to work is spread all over the language family.
Also, it basically ends up that every dependency gets into every `build` and `launch` image whether it needs it or not.
These buildpacks should be reworked to only require the correct versions of dependencies they need when they need them. 

- [x] https://github.com/paketo-buildpacks/dotnet-core-build/issues/85
- [x] https://github.com/paketo-buildpacks/dotnet-core-conf/issues/78

### Step 3 -- Stop requiring unnecessary dependencies in producer buildpacks

Currently, `dotnet-core-runtime`, `dotnet-core-aspnet`, and `dotnet-core-sdk` try to require themselves, and other dependencies, when they think other buildpacks in the language family will need them. This is an anti pattern, as the buildpacks that actually need them should require them. After step 2, `dotnet-core-conf` and `dotnet-core-build` should be requiring everything correctly, so these buildpacks can be reworked to only provide themselves.

- [x] https://github.com/paketo-buildpacks/dotnet-core-runtime/issues/84
- [x] https://github.com/paketo-buildpacks/dotnet-core-aspnet/issues/85
- [x] https://github.com/paketo-buildpacks/dotnet-core-sdk/issues/94

### Step 4 -- Restructure language family 

Up to now, the language family will still work because there is only one order group, so all dependencies will still show up on every image

Changing the `buildpack.toml` will make the final images contain only what they need.

- [x] A `build from source` order group.

The first order group should build `dotenet-core` apps from source. 

This group will trigger when no precompiled bits are provided.

```toml
[[order]] # Source

  [[order.group]]
    id = "paketo-buildpacks/dotnet-core-runtime"
    version = ""

  [[order.group]]
    id = "paketo-buildpacks/dotnet-core-aspnet"
    version = ""
    optional = true

  [[order.group]]
    id = "paketo-buildpacks/dotnet-core-sdk"
    version = ""

  [[order.group]]
    id = "paketo-buildpacks/icu"
    version = ""
    optional = true

  [[order.group]]
    id = "paketo-buildpacks/node-engine"
    version = ""
    optional = true

  [[order.group]]
    id = "paketo-buildpacks/dotnet-publish"
    version = ""

  [[order.group]]
    id = "paketo-buildpacks/dotnet-execute"
    version = ""
```

- [x] A `Framework-dependent deployment` and `Framework-dependent executable`

This group will trigger when the app source code contains `dll` or binary that needs framework library files in order to run

```toml
[[order]] # Framework-dependent deployment and Framework-dependent executable

  [[order.group]]
    id = "paketo-buildpacks/dotnet-core-runtime"
    version = ""

  [[order.group]]
    id = "paketo-buildpacks/dotnet-core-aspnet"
    version = ""
    optional = true

  [[order.group]]
    id = "paketo-buildpacks/icu"
    version = ""
    optional = true

  [[order.group]]
    id = "paketo-buildpacks/node-engine"
    version = ""
    optional = true

  [[order.group]]
    id = "paketo-buildpacks/dotnet-execute"
    version = ""
```

- [ ] A `self contained binary` order group

This group will trigger when the source code contains a binary file that has all framework libraries complied into it.

```toml
[[order]] # Self-contained deployment

  [[order.group]]
    id = "paketo-buildpacks/icu"
    version = ""
    optional = true

  [[order.group]]
    id = "paketo-buildpacks/node-engine"
    version = ""
    optional = true

  [[order.group]]
    id = "paketo-buildpacks/dotnet-execute"
    version = ""
```

**Comment by sophiewigmore** [comment]
@joshzarrabi  This looks great! One small thing I noticed is that a lot of the READMEs for the different dotnet repositories reference `dotnet-core-build` and `dotnet-core-conf`. See [ICU README](https://github.com/paketo-buildpacks/icu#icu-cloud-native-buildpack) and [Dotnet-core-sdk README](https://github.com/paketo-buildpacks/dotnet-core-sdk#dotnet-core-sdk-cloud-native-buildpack) for examples. Do you think we should add a task to go through and clean up references to dotnet-core-build/conf to dotnet-publish/execute so it doesn't get forgotten?

**Comment by joshzarrabi** [comment]
@sophiewigmore Yeah definitely, good catch.

**Comment by fg-j** [comment]
Note: I think the FDD/FDE order group needs to be updated to optionally include the SDK buildpack in cases where the `dotnet` CLI is needed at launch:
```
[[order]] # Framework-dependent deployment and Framework-dependent executable

  [[order.group]]
    id = "paketo-buildpacks/dotnet-core-runtime"
    version = ""

  [[order.group]]
    id = "paketo-buildpacks/dotnet-core-aspnet"
    version = ""
    optional = true

  [[order.group]]
    id = "paketo-buildpacks/dotnet-core-sdk"
    version = ""
    optional= true


  [[order.group]]
    id = "paketo-buildpacks/icu"
    version = ""
    optional = true

  [[order.group]]
    id = "paketo-buildpacks/node-engine"
    version = ""
    optional = true

  [[order.group]]
    id = "paketo-buildpacks/dotnet-execute"
    version = ""
```
Note that the dotnet execute buildpack [requires the SDK at launch if it detects the FDD case](https://github.com/paketo-buildpacks/dotnet-execute/blob/98f59e15ea74abe2991da5e0cc60f28ab59a4938/detect.go#L70)

**Comment by ryanmoran** [comment]
This makes sense. I believe we had previously assumed that the CLI was in the `runtime`.

**Comment by kvedurmu** [comment]
**Acceptance**
This looks great. I verified that ordering was updated as specified, downloaded the latest language family draft release artifact, and confirmed that the following app configurations could be built successfully:
- source based .NET runtime app
- self-contained app
- framework-dependent deployment
- framework-dependent executable

As a followup, I think it would be good to have a set of sample apps for all supported configurations. Filed - https://github.com/paketo-buildpacks/samples/issues/67 to track this.
